### ISSUE TITLE
TLS ECDH curve selection

### ISSUE THREAD

**Issue by adrien-n** [processone/ejabberd#1714]
Hi,

I am trying to make an ejabberd and a prosody server federate through TLS connections that use ECDH.

Currently prosody doesn't make it possible to allow several such curves (mostly because of a dependency) and defaults to secp384. As far as I can tell, ejabberd defaults to some 256 bits curve, possibly secp256.

I haven't found documentation about this in ejabberd so far and that makes me guess that it is impossible to chose another curve than the default one. As such, I am currently unable to federate across ejabberd and prosody servers.

I can think of several improvements to carry out:
- improve documentation (possibly merely by stating there is currently a limitation in this aspect),
- make this configurable,
- don't make this configurable (yet) but specify more curves in the default settings in order to have a non-empty intersection with other servers' configuration.

**Comment by zinid** [comment]
I think we need some motivation for this. We usually do not add options for special snowflakes who is resisting to use age-proven technologies like RSA and is creating compatibiliy problems for themselves and others.

**Comment by adrien-n** [comment]
I think I was a bit too deep in my thoughts and wasn't precise enough.
Sorry for that.

The problem is for federation: some servers use ciphersuites which
cannot be met with ejabberd and this makes federation impossible more
and more often.

The crypto itself is already supported but the list of curves used for
it is too restricted (only the one named "P-256" as far as I know). This
is an issue with configuration. It is similar to having one server only
do AES128 and the other only do AES256 (actually there are more
differences between AES options than between ECDH curves).

Also, this is strictly for the key exchange process of TLS, not for the
authentication or encryption parts. If it replaces something, it's the
non-elliptic curve Diffie Hellman ("DH") but not RSA or AES.

These things are not new and have been deployed and used in many places
for a fairly long time so why this ticket now?
The reason is that people are more and more moving to newer cipher
suites and the most modern ones all require ECDH. For instance, Mozilla
has a fairly good wiki page for server-side configuration of TLS and
three meaningful cipher lists: "modern", "intermediate" and "old".

  https://wiki.mozilla.org/Security/Server_Side_TLS#Recommended_configurations

In the "modern" one, all entries use ECDH(E). They provide other suites
for "services which need to support a wide range of clients (Firefox 1,
Chrome 1, IE 7, Opera 5 and Safari 1)". New server deployments don't try
to support these and go for the "modern" list.

Since this is TLS, the settings used for a given communication are
negotiated between both sides. They each send the cipher suites they can
use along with the parameters they handle and the best-ranked settings
are then used.

In each of the mozilla ciphersuites, they mention the curves available
for ECDH:
>  Certificate curve: prime256v1, secp384r1, secp521r1 

And that's where there is a need for a change: prime256v1 is also known
as P-256 and is used by ejabberd, secp384r1 is also known as P-384 and
is used by prosody. When a prosody and an ejabberd server want to talk,
they quite likely both have "ECDHE-RSA-AES256-GCM-SHA384" but the
ejabberd one only does ECDH with P-256 and the prosody one only with
P-384. This entirely removes the ability to do ECDH and to federate.

Prosody is going to be able to advertise multiple curves soonish but
this has to trickle down to distributions and installations. Similarly,
it will take some time for a change in ejabberd to be propagated
everywhere. By then, the current "compatibility" ciphersuites will be
old ones and ECDHE will not be avoidable.

I hope that clears any ambiguity about this need.

**Comment by adrien-n** [comment]
I was going through documentation for TLS 1.3 which is still a draft but which several implementors and groups have been pushing for quite strongly even if for HTTPS rather than XMPP.

TLS 1.3 will mandate the use of DH or ECDH. In addition to that, openssl has only added support for ECDH and *not* DH so far.

The fact is that the world is moving towards the use of ECDH so a good support is needed and that implies the ability to use several curves to be able to communicate with the various stacks that exist.

**Comment by mytbk** [comment]
Same problem to me.
I just found that if I only have "ECDHE+HIGH" in s2s_ciphers, then I cannot do s2s connections to many servers, just because we're using different curves. Also, I think it's weird that incoming s2s connections and outgoing s2s connections are using the same cipher options.
I hope ejabberd can give an option to set the ECDH curves. If I can set multiple curves, it'll be great, but I don't know if XMPP starttls supports this.

**Comment by adrien-n** [comment]
I've been able to spend more time on this issue.

Right now, federation is broken between ejabberd and prosody for quite a few deployments and there are going to be more and more problems because the versions in Debian Stretch can't communicate together and the issue is likely the same with other Linux distributions.

It also occured to me that ejabberd already supports everything except proposing several DH curves for the ECDH part. The ECDH part is already setup: it's only the curve choice that should be expanded. Hopefully that should be fairly easy to do and it doesn't even have to be very visible in the configuration because sane defaults are well-know nowadays: P-256, P-384 and P-512 (and they are the ones and only ones defined in the upcoming TLS 1.3 spec whereas there used to be many more possible but little-used).

In other words, default list should be P-256, P-384 and P-512, possibly with a setting to override this but I'm not even sure one is absolutely needed.

**Comment by adrien-n** [comment]
I just found this link: https://koldfront.dk/archive/2017/06/20-210822.html .

It mentions both #14 here and a patch against erlang-p1-tls. I can't say what the best place for changes is but these are certainly good to mention.

**Comment by zinid** [comment]
What makes me wondering is WHY an application developer should mess with this shit? Why wouldn't openssl provide sane defaults instead? Or provide broader curves list by default? Why wouldn't you guys create the issue in openssl bugtracker?
These crypto maniacs already created a system which is barely understandable (unless you spend several weeks of your life on the topic) and now its an application developers burden, great.

**Comment by adrien-n** [comment]
Well, I'm a user and it is ejabberd which is user-facing. It's not that
I am unwilling to locate what needs to be changed but that it is
difficult for me to do so because I am not familiar with Erlang nor with
the Erlang/OTP ecosystem or with the ejabberd code.

One thing that needs to be checked at least is that ejabberd uses
whatever current API should be used. This topic has evolved over the
past few years and while we're talkign about three possible values for
this, there used to be many many more that were available. What happened
is that only these three have seen widespread use but that couldn't be
foreseen half a decade ago.

A fairly simple thing which is difficult for me to check is whether
ejabberd explicitly uses P-256 a.k.a. secp256r1 or if it is a default
that has been inherited through library dependencies. The other aspect
is the one I mentioned above is to ensure that ejabberd is using APIs
that do not prevent newer options from being used because of
backward-compatibility.

I also have to stress the fact that federation is currently seriously
broken in practice. There are many things to take into account for
interoperability and no blame to put but the fact is that right now many
servers can't communicate together and this is only getting worse. In
order to restore this as quickly as possible, it is best to have all
components update (both ejabberd, prosody and other servers, and their
dependencies) rather than wait and hope for all prosody servers to
update. If it were only for me, I would have worked around the issue a
long time ago but there are many more people impacted by this.

**Comment by zinid** [comment]
You don't need to be an Erlang programmer, ejabberd is using openssl binding, [here it is](https://github.com/processone/fast_tls/blob/master/c_src/fast_tls.c), go figure what needs to be added.
Regarding federation and stuff. In my turn I'd like to stress, that the majority of ejabberd users are totally clueless and cannot even obtain a certificate for their domains. And we're talking about adding new TLS options, LOL? The defaults should be sane.

**Comment by pitchum** [comment]
@zinid I agree that ejabberd users shouldn't have to care about a new TLS option. Especially because it would probably break federation: we would end up with small groups of XMMP servers able to talk to each other, depending on which curve were chosen by their respective administrators.

Actually this is happening right now while Debian users are upgrading from jessie to stretch.
Prosody default configuration uses P-384 and ejabberd default configuration uses P-256.

I'm no TLS expert and I wonder why the choice of curve type (P-256, P-384 or P-512) is not part of the cipher negociation between 2 XMPP servers.
Do you think curve negociation is possible?

**Comment by zinid** [comment]
> I'm no TLS expert and I wonder why the choice of curve type (P-256, P-384 or P-512) is not part of the cipher negociation between 2 XMPP servers.

Me neither

> Do you think curve negociation is possible?

No idea

**Comment by adrien-n** [comment]
Curve negotiation is indeed possible: both sides send a list of curves rather than a single one.

Thanks for the link to the code in fast_tls. I had taken a wrong turn at some point and didn't end up looking at that file. The issue is indeed there and very obvious: https://github.com/processone/fast_tls/blob/master/c_src/fast_tls.c#L385 hardcodes P-256 (with another name but it's P-256). So I'm going to open a bug report in that repository.
